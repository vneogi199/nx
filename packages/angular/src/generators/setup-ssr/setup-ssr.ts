import type { Tree } from '@nx/devkit';
import {
  addDependenciesToPackageJson,
  formatFiles,
  installPackagesTask,
} from '@nx/devkit';
import {
  getInstalledPackageVersionInfo,
  versions,
} from '../utils/version-utils';
import {
  addHydration,
  generateSSRFiles,
  normalizeOptions,
  updateAppModule,
  updateProjectConfig,
  validateOptions,
} from './lib';
import type { Schema } from './schema';

export async function setupSsr(tree: Tree, schema: Schema) {
  validateOptions(tree, schema);
  const options = normalizeOptions(tree, schema);

  updateProjectConfig(tree, options);
  generateSSRFiles(tree, options);

  if (!options.standalone) {
    updateAppModule(tree, options);
  }

  if (options.hydration) {
    addHydration(tree, options);
  }

  const pkgVersions = versions(tree);

  addDependenciesToPackageJson(
    tree,
    {
      '@nguniversal/express-engine':
        getInstalledPackageVersionInfo(tree, '@nguniversal/express-engine')
          ?.version ?? pkgVersions.ngUniversalVersion,
      '@angular/platform-server':
        getInstalledPackageVersionInfo(tree, '@angular/platform-server')
          ?.version ?? pkgVersions.angularVersion,
    },
    {
      '@nguniversal/builders':
        getInstalledPackageVersionInfo(tree, '@nguniversal/builders')
          ?.version ?? pkgVersions.ngUniversalVersion,
    }
  );

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return () => {
    installPackagesTask(tree);
  };
}

export default setupSsr;
