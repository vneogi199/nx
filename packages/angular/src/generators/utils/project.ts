import type { Tree } from '@nx/devkit';
import { readProjectConfiguration } from '@nx/devkit';
import type { AngularProjectConfiguration } from '../../utils/types';
import { getNpmScope } from '@nx/js/src/utils/package-json/get-npm-scope';

export function normalizeNewProjectPrefix(
  prefix: string | undefined,
  npmScope: string | undefined,
  fallbackPrefix: string
): string {
  // Prefix needs to be a valid html selector, if npmScope it's not valid, we don't default
  // to it and let it fall through to the Angular schematic to handle it
  // https://github.com/angular/angular-cli/blob/1c634cd327e5a850553b258aa2d5e6a6b2c75c65/packages/schematics/angular/component/index.ts#L130
  const htmlSelectorRegex =
    /^[a-zA-Z][.0-9a-zA-Z]*(:?-[a-zA-Z][.0-9a-zA-Z]*)*$/;

  if (prefix) {
    if (!htmlSelectorRegex.test(prefix)) {
      throw new Error(
        'The provided "prefix" is invalid. The prefix must start with a letter, and must contain only alphanumeric characters or dashes. When adding a dash the segment after the dash must also start with a letter.'
      );
    }

    return prefix;
  }

  if (npmScope && !htmlSelectorRegex.test(npmScope)) {
    throw new Error(`The "--prefix" option was not provided, therefore attempted to use the "npmScope" defined in "nx.json" to set the application's selector prefix, but it is invalid.

There are two options that can be followed to resolve this issue:
  - Pass a valid "--prefix" option.
  - Update the "npmScope" in "nx.json" (Note: this can be an involved process, as other libraries and applications may need to be updated to match the new scope).

If you encountered this error when creating a new Nx Workspace, the workspace name or "npmScope" is invalid to use as the selector prefix for the application being generated.

Valid selector prefixes must start with a letter, and must contain only alphanumeric characters or dashes. When adding a dash the segment after the dash must also start with a letter.`);
  }

  return npmScope || fallbackPrefix;
}

export function getProjectPrefix(
  tree: Tree,
  project: string
): string | undefined {
  return (
    (readProjectConfiguration(tree, project) as AngularProjectConfiguration)
      .prefix ?? getNpmScope(tree)
  );
}
