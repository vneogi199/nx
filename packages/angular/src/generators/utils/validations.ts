import type { Tree } from '@nx/devkit';
import { getProjects, stripIndents } from '@nx/devkit';
import { lt } from 'semver';
import { getInstalledAngularVersionInfo } from './version-utils';

export function validateProject(tree: Tree, projectName: string): void {
  const projects = getProjects(tree);

  if (!projects.has(projectName)) {
    throw new Error(
      `Project "${projectName}" does not exist! Please provide an existing project name.`
    );
  }
}

export function validateStandaloneOption(
  tree: Tree,
  standalone: boolean | undefined,
  angularVersion?: string
): void {
  if (!standalone) {
    return;
  }

  const installedAngularVersion =
    angularVersion ?? getInstalledAngularVersionInfo(tree).version;

  if (lt(installedAngularVersion, '14.1.0')) {
    throw new Error(stripIndents`The "standalone" option is only supported in Angular >= 14.1.0. You are currently using "${installedAngularVersion}".
    You can resolve this error by removing the "standalone" option or by migrating to Angular 14.1.0.`);
  }
}
