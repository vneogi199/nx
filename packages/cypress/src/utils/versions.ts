export const nxVersion = require('../../package.json').version;
export const eslintPluginCypressVersion = '^2.13.4';
export const typesNodeVersion = '16.11.7';
export const cypressViteDevServerVersion = '^2.2.1';
export const cypressVersion = '^13.0.0';
export const cypressWebpackVersion = '^2.0.0';
export const webpackHttpPluginVersion = '^5.5.0';
export const viteVersion = '~4.3.9';
export const htmlWebpackPluginVersion = '^5.5.0';
