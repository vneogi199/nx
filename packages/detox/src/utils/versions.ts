export const nxVersion = require('../../package.json').version;

export const detoxVersion = '^20.11.1';
export const testingLibraryJestDom = '5.16.5';
export const configPluginsDetoxVersion = '~6.0.0'; // only required for expo
