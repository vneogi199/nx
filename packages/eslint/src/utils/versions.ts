export const nxVersion = require('../../package.json').version;

export const eslintVersion = '~8.46.0';
export const eslintrcVersion = '^2.1.1';
export const eslintConfigPrettierVersion = '^9.0.0';
export const typescriptESLintVersion = '^5.60.1';
