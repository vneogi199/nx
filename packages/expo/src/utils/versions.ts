export const nxVersion = require('../../package.json').version;

export const expoVersion = '49.0.10';
export const expoMetroConfigVersion = '~0.10.7';
export const expoSplashScreenVersion = '~0.20.5';
export const expoStatusBarVersion = '~1.6.0';
export const expoCliVersion = '~0.10.12'; // @expo/cli
export const easCliVersion = '~5.2.0';
export const babelPresetExpoVersion = '~9.5.2';

export const reactVersion = '18.2.0';
export const reactDomVersion = '18.2.0';
export const reactTestRendererVersion = '18.2.0';
export const typesReactVersion = '18.0.28';

export const reactNativeVersion = '0.72.4';
export const reactNativeWebVersion = '~0.19.8';

export const reactNativeSvgTransformerVersion = '1.0.0';
export const reactNativeSvgVersion = '13.9.0';

export const metroVersion = '0.76.7';

export const testingLibraryReactNativeVersion = '~12.3.0';
export const testingLibraryJestNativeVersion = '~5.4.3';
export const jestExpoVersion = '~49.0.0';
