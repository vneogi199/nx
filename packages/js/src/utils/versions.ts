export const nxVersion = require('../../package.json').version;

export const esbuildVersion = '^0.19.2';
export const prettierVersion = '^2.6.2';
export const swcCliVersion = '~0.1.62';
export const swcCoreVersion = '~1.3.85';
export const swcHelpersVersion = '~0.5.2';
export const swcNodeVersion = '~1.6.7';
export const tsLibVersion = '^2.3.0';
export const typesNodeVersion = '18.7.1';
export const verdaccioVersion = '^5.0.4';

// Typescript
export const typescriptVersion = '~5.1.3';
/**
 * The minimum version is currently determined from the lowest version
 * that's supported by the lowest Angular supported version, e.g.
 * `npm view @angular/compiler-cli@14.0.0 peerDependencies.typescript`
 */
export const supportedTypescriptVersions = '>=4.6.2';
