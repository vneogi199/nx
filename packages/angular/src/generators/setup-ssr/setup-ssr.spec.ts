import {
  NxJsonConfiguration,
  readJson,
  readProjectConfiguration,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { PackageJson } from 'nx/src/utils/package-json';
import { angularVersion, ngUniversalVersion } from '../../utils/versions';
import { generateTestApplication } from '../utils/testing';
import { setupSsr } from './setup-ssr';

describe('setupSSR', () => {
  it('should create the files correctly for ssr', async () => {
    // ARRANGE
    const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    await generateTestApplication(tree, {
      name: 'app1',
    });

    // ACT
    await setupSsr(tree, { project: 'app1' });

    // ASSERT
    expect(
      readProjectConfiguration(tree, 'app1').targets.server
    ).toMatchSnapshot();
    expect(tree.read('app1/server.ts', 'utf-8')).toMatchSnapshot();
    expect(tree.read('app1/src/main.server.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "export { AppServerModule } from './app/app.server.module';
      "
    `);
    expect(tree.read('app1/src/main.ts', 'utf-8')).toMatchInlineSnapshot(`
      "import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
      import { AppModule } from './app/app.module';

      platformBrowserDynamic()
        .bootstrapModule(AppModule)
        .catch((err) => console.error(err));
      "
    `);
    expect(tree.read('app1/tsconfig.server.json', 'utf-8'))
      .toMatchInlineSnapshot(`
      "/* To learn more about this file see: https://angular.io/config/tsconfig. */
      {
        "extends": "./tsconfig.app.json",
        "compilerOptions": {
          "outDir": "../../out-tsc/server",
          "target": "es2019",
          "types": ["node"]
        },
        "files": ["src/main.server.ts", "server.ts"]
      }
      "
    `);
    expect(tree.read('app1/src/app/app.server.module.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { NgModule } from '@angular/core';
      import { ServerModule } from '@angular/platform-server';

      import { AppModule } from './app.module';
      import { AppComponent } from './app.component';

      @NgModule({
        imports: [AppModule, ServerModule],
        bootstrap: [AppComponent],
      })
      export class AppServerModule {}
      "
    `);
    expect(tree.read('app1/src/app/app.module.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { NgModule } from '@angular/core';
      import { BrowserModule } from '@angular/platform-browser';
      import { AppComponent } from './app.component';
      import { NxWelcomeComponent } from './nx-welcome.component';

      @NgModule({
        declarations: [AppComponent, NxWelcomeComponent],
        imports: [BrowserModule],
        providers: [],
        bootstrap: [AppComponent],
      })
      export class AppModule {}
      "
    `);
    const packageJson = readJson<PackageJson>(tree, 'package.json');
    const dependencies = {
      '@nguniversal/express-engine': ngUniversalVersion,
      '@angular/platform-server': angularVersion,
    };
    for (const [dep, version] of Object.entries(dependencies)) {
      expect(packageJson.dependencies[dep]).toEqual(version);
    }
    const devDeps = {
      '@nguniversal/builders': ngUniversalVersion,
    };
    for (const [dep, version] of Object.entries(devDeps)) {
      expect(packageJson.devDependencies[dep]).toEqual(version);
    }
    const nxJson = readJson<NxJsonConfiguration>(tree, 'nx.json');
    expect(nxJson.targetDefaults.server.cache).toBe(true);
  });

  it('should use fileReplacements if they already exist', async () => {
    // ARRANGE
    const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    await generateTestApplication(tree, {
      name: 'app1',
    });

    tree.write('app1/src/environments/environment.ts', '');
    tree.write('app1/src/environments/environment.prod.ts', '');
    const project = readProjectConfiguration(tree, 'app1');
    project.targets.build.configurations.production.fileReplacements = [
      {
        replace: 'app1/src/environments/environment.ts',
        with: 'app1/src/environments/environment.prod.ts',
      },
    ];
    updateProjectConfiguration(tree, 'app1', project);

    // ACT
    await setupSsr(tree, { project: 'app1' });

    // ASSERT
    expect(
      readProjectConfiguration(tree, 'app1').targets.server
    ).toMatchSnapshot();
  });

  it('should create the files correctly for ssr when app is standalone', async () => {
    // ARRANGE
    const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    await generateTestApplication(tree, {
      name: 'app1',
      standalone: true,
    });

    // ACT
    await setupSsr(tree, { project: 'app1' });

    // ASSERT
    expect(
      readProjectConfiguration(tree, 'app1').targets.server
    ).toMatchSnapshot();
    expect(tree.read('app1/server.ts', 'utf-8')).toMatchSnapshot();
    expect(tree.read('app1/src/main.server.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { bootstrapApplication } from '@angular/platform-browser';
      import { AppComponent } from './app/app.component';
      import { config } from './app/app.config.server';

      const bootstrap = () => bootstrapApplication(AppComponent, config);

      export default bootstrap;
      "
    `);
    expect(tree.read('app1/tsconfig.server.json', 'utf-8'))
      .toMatchInlineSnapshot(`
      "/* To learn more about this file see: https://angular.io/config/tsconfig. */
      {
        "extends": "./tsconfig.app.json",
        "compilerOptions": {
          "outDir": "../../out-tsc/server",
          "target": "es2019",
          "types": ["node"]
        },
        "files": ["src/main.server.ts", "server.ts"]
      }
      "
    `);
    expect(tree.read('app1/src/app/app.config.server.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { mergeApplicationConfig, ApplicationConfig } from '@angular/core';
      import { provideServerRendering } from '@angular/platform-server';
      import { appConfig } from './app.config';

      const serverConfig: ApplicationConfig = {
        providers: [provideServerRendering()],
      };

      export const config = mergeApplicationConfig(appConfig, serverConfig);
      "
    `);

    const packageJson = readJson<PackageJson>(tree, 'package.json');
    const dependencies = {
      '@nguniversal/express-engine': ngUniversalVersion,
      '@angular/platform-server': angularVersion,
    };
    for (const [dep, version] of Object.entries(dependencies)) {
      expect(packageJson.dependencies[dep]).toEqual(version);
    }
    const devDeps = {
      '@nguniversal/builders': ngUniversalVersion,
    };
    for (const [dep, version] of Object.entries(devDeps)) {
      expect(packageJson.devDependencies[dep]).toEqual(version);
    }
    const nxJson = readJson<NxJsonConfiguration>(tree, 'nx.json');
    expect(nxJson.targetDefaults.server.cache).toEqual(true);
  });

  it('should add hydration correctly for NgModule apps', async () => {
    // ARRANGE
    const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    await generateTestApplication(tree, {
      name: 'app1',
    });

    // ACT
    await setupSsr(tree, { project: 'app1', hydration: true });

    // ASSERT
    expect(tree.read('app1/src/app/app.module.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { NgModule } from '@angular/core';
      import {
        BrowserModule,
        provideClientHydration,
      } from '@angular/platform-browser';
      import { AppComponent } from './app.component';
      import { NxWelcomeComponent } from './nx-welcome.component';

      @NgModule({
        declarations: [AppComponent, NxWelcomeComponent],
        imports: [BrowserModule],
        providers: [provideClientHydration()],
        bootstrap: [AppComponent],
      })
      export class AppModule {}
      "
    `);
  });

  it('should add hydration correctly to standalone', async () => {
    // ARRANGE
    const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    await generateTestApplication(tree, {
      name: 'app1',
      standalone: true,
    });

    // ACT
    await setupSsr(tree, { project: 'app1', hydration: true });

    // ASSERT
    expect(tree.read('app1/src/app/app.config.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { ApplicationConfig } from '@angular/core';
      import { provideClientHydration } from '@angular/platform-browser';

      export const appConfig: ApplicationConfig = {
        providers: [provideClientHydration()],
      };
      "
    `);

    expect(tree.read('app1/src/app/app.config.server.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { mergeApplicationConfig, ApplicationConfig } from '@angular/core';
      import { provideServerRendering } from '@angular/platform-server';
      import { appConfig } from './app.config';

      const serverConfig: ApplicationConfig = {
        providers: [provideServerRendering()],
      };

      export const config = mergeApplicationConfig(appConfig, serverConfig);
      "
    `);
  });

  describe('compat', () => {
    it('should install the correct versions when using older versions of Angular', async () => {
      // ARRANGE
      const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

      await generateTestApplication(tree, {
        name: 'app1',
      });

      updateJson(tree, 'package.json', (json) => ({
        ...json,
        dependencies: {
          '@angular/core': '14.2.0',
        },
      }));

      // ACT
      await setupSsr(tree, { project: 'app1' });

      // ASSERT
      const pkgJson = readJson(tree, 'package.json');
      expect(pkgJson.dependencies['@angular/platform-server']).toEqual(
        '~14.2.0'
      );
      expect(pkgJson.dependencies['@nguniversal/express-engine']).toEqual(
        '~14.2.0'
      );
      expect(pkgJson.devDependencies['@nguniversal/builders']).toEqual(
        '~14.2.0'
      );
    });

    it('should create the main.server.ts file correctly for Angular v14', async () => {
      // ARRANGE
      const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
      await generateTestApplication(tree, {
        name: 'app1',
      });
      updateJson(tree, 'package.json', (json) => ({
        ...json,
        dependencies: { '@angular/core': '14.2.0' },
      }));

      // ACT
      await setupSsr(tree, { project: 'app1' });

      // ASSERT
      expect(tree.read('app1/src/main.server.ts', 'utf-8'))
        .toMatchInlineSnapshot(`
        "/***************************************************************************************************
         * Initialize the server environment - for example, adding DOM built-in types to the global scope.
         *
         * NOTE:
         * This import must come before any imports (direct or transitive) that rely on DOM built-ins being
         * available, such as \`@angular/elements\`.
         */
        import '@angular/platform-server/init';

        export { AppServerModule } from './app/app.server.module';
        export { renderModule } from '@angular/platform-server';
        "
      `);
    });

    it('should add "withServerTransition" call to app module for angular versions lower than 16', async () => {
      // ARRANGE
      const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
      updateJson(tree, 'package.json', (json) => ({
        ...json,
        dependencies: { ...json.dependencies, '@angular/core': '^15.2.0' },
      }));

      await generateTestApplication(tree, {
        name: 'app1',
      });

      // ACT
      await setupSsr(tree, { project: 'app1' });

      // ASSERT
      expect(tree.read('app1/src/app/app.module.ts', 'utf-8'))
        .toMatchInlineSnapshot(`
              "import { NgModule } from '@angular/core';
              import { BrowserModule } from '@angular/platform-browser';
              import { AppComponent } from './app.component';
              import { NxWelcomeComponent } from './nx-welcome.component';

              @NgModule({
                declarations: [AppComponent, NxWelcomeComponent],
                imports: [BrowserModule.withServerTransition({ appId: 'serverApp' })],
                providers: [],
                bootstrap: [AppComponent],
              })
              export class AppModule {}
              "
          `);
    });

    it('should wrap bootstrap call for Angular versions lower than 15.2', async () => {
      // ARRANGE
      const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
      await generateTestApplication(tree, {
        name: 'app1',
      });
      updateJson(tree, 'package.json', (json) => ({
        ...json,
        dependencies: { '@angular/core': '15.1.0' },
      }));

      // ACT
      await setupSsr(tree, { project: 'app1' });

      // ASSERT
      expect(tree.read('app1/src/main.ts', 'utf-8')).toMatchInlineSnapshot(`
              "import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
              import { AppModule } from './app/app.module';

              function bootstrap() {
                platformBrowserDynamic()
                  .bootstrapModule(AppModule)
                  .catch((err) => console.error(err));
              }

              if (document.readyState !== 'loading') {
                bootstrap();
              } else {
                document.addEventListener('DOMContentLoaded', bootstrap);
              }
              "
          `);
    });
  });
});
