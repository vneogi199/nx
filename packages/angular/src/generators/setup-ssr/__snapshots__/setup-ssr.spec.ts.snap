// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupSSR should create the files correctly for ssr 1`] = `
{
  "configurations": {
    "development": {
      "buildOptimizer": false,
      "extractLicenses": false,
      "optimization": false,
      "sourceMap": true,
      "vendorChunk": true,
    },
    "production": {
      "outputHashing": "media",
    },
  },
  "defaultConfiguration": "production",
  "dependsOn": [
    "build",
  ],
  "executor": "@angular-devkit/build-angular:server",
  "options": {
    "main": "app1/server.ts",
    "outputPath": "dist/app1/server",
    "tsConfig": "app1/tsconfig.server.json",
  },
}
`;

exports[`setupSSR should create the files correctly for ssr 2`] = `
"import 'zone.js/dist/zone-node';

import { APP_BASE_HREF } from '@angular/common';
import { ngExpressEngine } from '@nguniversal/express-engine';
import * as express from 'express';
import { existsSync } from 'fs';
import { join } from 'path';

import { AppServerModule } from './src/main.server';

// The Express app is exported so that it can be used by serverless Functions.
export function app(): express.Express {
  const server = express();
  const distFolder = join(process.cwd(), 'dist/app1/browser');
  const indexHtml = existsSync(join(distFolder, 'index.original.html'))
    ? 'index.original.html'
    : 'index';

  // Our Universal express-engine (found @ https://github.com/angular/universal/tree/main/modules/express-engine)
  server.engine(
    'html',
    ngExpressEngine({
      bootstrap: AppServerModule,
    })
  );

  server.set('view engine', 'html');
  server.set('views', distFolder);

  // Example Express Rest API endpoints
  // server.get('/api/**', (req, res) => { });
  // Serve static files from /browser
  server.get(
    '*.*',
    express.static(distFolder, {
      maxAge: '1y',
    })
  );

  // All regular routes use the Universal engine
  server.get('*', (req, res) => {
    res.render(indexHtml, {
      req,
      providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
  });

  return server;
}

function run(): void {
  const port = process.env['PORT'] || 4000;

  // Start up the Node server
  const server = app();
  server.listen(port, () => {
    console.log(\`Node Express server listening on http://localhost:\${port}\`);
  });
}

// Webpack will replace 'require' with '__webpack_require__'
// '__non_webpack_require__' is a proxy to Node 'require'
// The below code is to ensure that the server is run only when not requiring the bundle.
declare const __non_webpack_require__: NodeRequire;
const mainModule = __non_webpack_require__.main;
const moduleFilename = (mainModule && mainModule.filename) || '';
if (moduleFilename === __filename || moduleFilename.includes('iisnode')) {
  run();
}

export * from './src/main.server';
"
`;

exports[`setupSSR should create the files correctly for ssr when app is standalone 1`] = `
{
  "configurations": {
    "development": {
      "buildOptimizer": false,
      "extractLicenses": false,
      "optimization": false,
      "sourceMap": true,
      "vendorChunk": true,
    },
    "production": {
      "outputHashing": "media",
    },
  },
  "defaultConfiguration": "production",
  "dependsOn": [
    "build",
  ],
  "executor": "@angular-devkit/build-angular:server",
  "options": {
    "main": "app1/server.ts",
    "outputPath": "dist/app1/server",
    "tsConfig": "app1/tsconfig.server.json",
  },
}
`;

exports[`setupSSR should create the files correctly for ssr when app is standalone 2`] = `
"import 'zone.js/dist/zone-node';

import { APP_BASE_HREF } from '@angular/common';
import { ngExpressEngine } from '@nguniversal/express-engine';
import * as express from 'express';
import { existsSync } from 'fs';
import { join } from 'path';

import bootstrap from './src/main.server';

// The Express app is exported so that it can be used by serverless Functions.
export function app(): express.Express {
  const server = express();
  const distFolder = join(process.cwd(), 'dist/app1/browser');
  const indexHtml = existsSync(join(distFolder, 'index.original.html'))
    ? 'index.original.html'
    : 'index';

  // Our Universal express-engine (found @ https://github.com/angular/universal/tree/main/modules/express-engine)
  server.engine(
    'html',
    ngExpressEngine({
      bootstrap,
    })
  );

  server.set('view engine', 'html');
  server.set('views', distFolder);

  // Example Express Rest API endpoints
  // server.get('/api/**', (req, res) => { });
  // Serve static files from /browser
  server.get(
    '*.*',
    express.static(distFolder, {
      maxAge: '1y',
    })
  );

  // All regular routes use the Universal engine
  server.get('*', (req, res) => {
    res.render(indexHtml, {
      req,
      providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
  });

  return server;
}

function run(): void {
  const port = process.env['PORT'] || 4000;

  // Start up the Node server
  const server = app();
  server.listen(port, () => {
    console.log(\`Node Express server listening on http://localhost:\${port}\`);
  });
}

// Webpack will replace 'require' with '__webpack_require__'
// '__non_webpack_require__' is a proxy to Node 'require'
// The below code is to ensure that the server is run only when not requiring the bundle.
declare const __non_webpack_require__: NodeRequire;
const mainModule = __non_webpack_require__.main;
const moduleFilename = (mainModule && mainModule.filename) || '';
if (moduleFilename === __filename || moduleFilename.includes('iisnode')) {
  run();
}

export default bootstrap;
"
`;

exports[`setupSSR should use fileReplacements if they already exist 1`] = `
{
  "configurations": {
    "development": {
      "buildOptimizer": false,
      "extractLicenses": false,
      "optimization": false,
      "sourceMap": true,
      "vendorChunk": true,
    },
    "production": {
      "fileReplacements": [
        {
          "replace": "app1/src/environments/environment.ts",
          "with": "app1/src/environments/environment.prod.ts",
        },
      ],
      "outputHashing": "media",
    },
  },
  "defaultConfiguration": "production",
  "dependsOn": [
    "build",
  ],
  "executor": "@angular-devkit/build-angular:server",
  "options": {
    "main": "app1/server.ts",
    "outputPath": "dist/app1/server",
    "tsConfig": "app1/tsconfig.server.json",
  },
}
`;
